src/Type/Doctrine/DBAL/QueryBuilder/QueryBuilderExecuteMethodExtension.php

```diff
@@ -0,0 +1,51 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Doctrine\DBAL\QueryBuilder;
+
+use Doctrine\DBAL\Driver\ResultStatement;
+use Doctrine\DBAL\Query\QueryBuilder;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Identifier;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\DynamicMethodReturnTypeExtension;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+
+class QueryBuilderExecuteMethodExtension implements DynamicMethodReturnTypeExtension
+{
+
+	public function getClass(): string
+	{
+		return QueryBuilder::class;
+	}
+
+	public function isMethodSupported(MethodReflection $methodReflection): bool
+	{
+		return $methodReflection->getName() === 'execute';
+	}
+
+	public function getTypeFromMethodCall(MethodReflection $methodReflection, MethodCall $methodCall, Scope $scope): Type
+	{
+		$defaultReturnType = ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+
+		$name = '';
+		$var = $methodCall->var;
+		while ($var instanceof MethodCall) {
+			$nameObject = $var->name;
```

Comment:
this could be `Identifier|Expr` so I need to check on the correct type to get the name string. I cannot simply cast them to string. without this if below, PHPStan would complain :wink:

Reply:
Yes, this is fine, in the else branch I guess we could return `$defaultReturnType` instead of continuing with the loop.

Reply:
I used an early return. hope I got that right :blush: